include/dynamatic/Dialect/Handshake/HandshakeOps.td

```diff
@@ -596,20 +604,29 @@ def MemoryControllerOp : Handshake_Op<"mem_controller", [
     3. Control signal indicating completion.
     ```
   }];
-  
-  let arguments = (ins AnyMemRef:$memRef, ControlType:$memStart,
-                       Variadic<ChannelType>:$inputs, ControlType:$ctrlEnd,
+
+  // The memory controller does not accept ControlTypes with extra signals.
+  // Handling extra signals is the responsibility of MemPortOps (e.g., LoadOp, StoreOp, ...).
+  // (at least in speculative execution scenarios)
+  // TODO: Change ChannelType to SimpleChannel as well
+  let arguments = (ins AnyMemRef:$memRef, SimpleControl:$memStart,
+                       Variadic<ChannelType>:$inputs, SimpleControl:$ctrlEnd,
```

Comment:
I think we open an issue about the memory controller and tag both Lucas and Aya? before changing ChannelType to SimpleChannel anyway

Reply:
we can discuss exactly what the comment should look like after we talk with them

Reply:
Thank you for opening #214 - we'll decide this after discussing it

Reply:
Yeah you can make all of these argument/result types simple.

Reply:
I found the `Variadic` can only accept a (pure) Type, not TypeConstraint. I'll come up with the solution later.
I'd be glad if you know something already and let me know it

Reply:
Interesting, it might not be possible. The solution is to manually verify that the relevant results/operands are simple types in the op's verification method.

Reply:
@shundroid  Doing something else and noticed it is in fact possible.

https://github.com/EPFL-LAP/dynamatic/blob/86e50729d5ed52561f9fe3e3011a95094138d6b8/include/dynamatic/Dialect/Handshake/HandshakeOps.td#L1090-L1092

`SignalType` is a `Type` with an additional custom constraint.

Reply:
SignalType is a `Type`.
https://github.com/EPFL-LAP/dynamatic/blob/86e50729d5ed52561f9fe3e3011a95094138d6b8/include/dynamatic/Dialect/Handshake/HandshakeTypes.td#L145-L148
But ChannelType is a `TypeConstraint`.
https://github.com/EPFL-LAP/dynamatic/blob/86e50729d5ed52561f9fe3e3011a95094138d6b8/include/dynamatic/Dialect/Handshake/HandshakeTypes.td#L184-L188

And the reason for this error is that Variadic only accepts `Type`, not `TypeConstraint`:
![image](https://github.com/user-attachments/assets/a7d98f79-9c74-4b4d-9973-263131d67d88)

I'm not sure why SignalType is a Type whereas ChannelType is a TypeConstraint.
Is it okay to change ChannelType to a "Type with constraints"?

Reply:
Yes I meant that it is possible to put a `Type` there that has some verification attached, not a literal `TypeConstraint`.

I think I used `TypeConstrait` for SignalType to compose it with other constraints. Yes you can change `SimpleChannel` to a "type with constraints" if it allows your use case and does not break other existing ones.

Reply:
Oh sorry, reading your comment again I realized you were suggesting exactly this.
I changed SimpleChannel/Control to Type (from TypeConstraint) and it worked! Thanks!